src/Metadata/Resource/Factory/XmlResourceMetadataFactory.php

```diff
@@ -0,0 +1,206 @@
+                    continue;
+                }
+
+                $metadata = $this->toArray($resource);
+
+                foreach ($resource->attributes as $attribute) {
+                    $metadata[$attribute->name] = $attribute->value;
+                }
+
+                break;
+            }
+
+            if (null !== $metadata) {
+                break;
+            }
+        }
+
+        if (!$metadata || count($metadata) === 0) {
+            return $this->handleNotFound($parentResourceMetadata, $resourceClass);
+        }
+
+        if (!$parentResourceMetadata) {
+            return new ResourceMetadata(
+                $metadata['shortName'] ?? null,
+                $metadata['description'] ?? null,
+                $metadata['iri'] ?? null,
+                $metadata['itemOperations'] ?? null,
+                $metadata['collectionOperations'] ?? null,
+                $metadata['attributes'] ?? []
+            );
+        }
+
+        $resourceMetadata = $parentResourceMetadata;
+
+        foreach (['shortName', 'description', 'itemOperations', 'collectionOperations', 'iri', 'attributes'] as $property) {
+            if (!isset($metadata[$property])) {
+                continue;
+            }
+
+            $resourceMetadata = $this->createWith($resourceMetadata, $property, $metadata[$property]);
+        }
+
+        return $resourceMetadata;
+    }
+
+    /**
+     * Returns the metadata from the decorated factory if available or throws an exception.
+     *
+     * @param ResourceMetadata|null $parentPropertyMetadata
+     * @param string                $resourceClass
+     *
+     * @throws ResourceClassNotFoundException
+     *
+     * @return ResourceMetadata
+     */
+    private function handleNotFound(ResourceMetadata $parentPropertyMetadata = null, string $resourceClass) : ResourceMetadata
+    {
+        if (null !== $parentPropertyMetadata) {
+            return $parentPropertyMetadata;
+        }
+
+        throw new ResourceClassNotFoundException(sprintf('Resource "%s" not found.', $resourceClass));
+    }
+
+    /**
+     * Creates a new instance of metadata if the property is not already set.
+     *
+     * @param ResourceMetadata $resourceMetadata
+     * @param string           $property
+     * @param mixed            $value
+     *
+     * @return ResourceMetadata
+     */
+    private function createWith(ResourceMetadata $resourceMetadata, string $property, $value) : ResourceMetadata
+    {
+        $getter = 'get'.ucfirst($property);
+
+        if (null !== $resourceMetadata->$getter()) {
+            return $resourceMetadata;
+        }
+
+        $wither = 'with'.ucfirst($property);
+
+        return $resourceMetadata->$wither($value);
+    }
+
+    /**
+     * Transforms an xml dom to an array.
+     *
+     * @param DomElement                    $node
+     * @param nodes relative root reference $nodes
+     */
+    private function toArray(DomElement $node, array &$nodes = []) : array
```

Comment:
Is it really necessary to convert the XML to an intermediate representation? IMO it would be easier to maintain if we create directly the metadata here.